plumbing/format/pktline/reader.go

```diff
@@ -0,0 +1,192 @@
+package pktline
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/go-git/go-git/v5/utils/trace"
+)
+
+var (
+	// ErrNegativeCount is returned by Read when the count is negative.
+	ErrNegativeCount = errors.New("negative count")
+)
+
+// Reader represents a pktline reader.
+type Reader struct {
+	r io.Reader
+
+	buf []byte // peeked buffer
+}
+
+// NewReader returns a new pktline reader that reads from r and supports
+// peeking.
+func NewReader(r io.Reader) *Reader {
+	if rdr, ok := r.(*Reader); ok {
```

Comment:
Should NewReader return the io.Reader interface? Since the reader struct implements the interface. But we'll have a different architecture in this case. Makes sense?